### PR TITLE
Add IR Noblex climate component

### PR DESCRIPTION
**Pull request by AGalfra**
# What does this implement/fix?

This adds a new climate component to the ESPHome ecosystem, providing compatibility with Noblex brand climate equipment.
This code is available as an external component from my github: https://github.com/AGalfra/noblex.git

## Types of changes

- New feature

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation:** esphome/esphome-docs#3061

## Test Environment

-  ESP32
- ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
climate:
  - platform: noblex
    name: "Prueba AC"
    id: prueba_ac
    sensor: temp
    receiver_id: rcvr
```

## Checklist:
  - The code change is tested and works locally.
  - Tests have been added to verify that the new code works (under `tests/test1.yaml` folder).

### CODE REVIEW

**Comment by AGalfra** [comment]
Hello.
This is my first contribution to ESPHome.
I've been using it for a long time and it's time to contribute to the project.
thank you so much.

**Comment by AGalfra** [comment]
hi @jesserockz could you verify this pull request?

**Comment by AGalfra** [comment]
Hi dear @jesserockz now the documentation is ready https://github.com/esphome/esphome-docs/pull/3054.

File: esphome/components/noblex/noblex.cpp [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+uint8_t NoblexClimate::swap_(uint8_t num) {
+  uint8_t value = 0;
+  for (int8_t i = 7; i >= 0; i--) {
+    if (num & (1 << (7 - i))) {
+      value |= (1 << i);
+    }
+  }
+  return value;
+}
```
Comment (jesserockz) [inline esphome/components/noblex/noblex.cpp]: ESPHome has `uint8_t reverse_bits(uint8_t)` in `helpers.h` for this already.

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/noblex/noblex.cpp (hunk 1 above)
Comment (AGalfra) [reply esphome/components/noblex/noblex.cpp]: thank you for the correction. I already modified the file and I am using the function included in helpers

Comment (AGalfra) [reply esphome/components/noblex/noblex.cpp]: @jesserockz.
dear, the suggested correction is ready

File: esphome/components/noblex/noblex.cpp [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+      break;
+    case climate::CLIMATE_SWING_OFF:
+    default:
+      remote_state[0] &= 0xFD;
+      remote_state[4] = 0x50;
+      break;
+  }
+
+  uint8_t crc = 0;
+  for (unsigned char i : remote_state) {
```
Comment (jesserockz) [inline esphome/components/noblex/noblex.cpp]:
```suggestion
  for (uint8_t i : remote_state) {
```

File: esphome/components/noblex/noblex.cpp [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+    // Validate footer
+    if (!data.expect_mark(NOBLEX_BIT_MARK)) {
+      ESP_LOGV(TAG, "Footer fail");
+      return false;
+    }
+    receiving = false;
+  }
+
+  for (unsigned char i : remote_state)
```
Comment (jesserockz) [inline esphome/components/noblex/noblex.cpp]:
```suggestion
  for (uint8_t i : remote_state)
```

File: esphome/components/noblex/noblex.cpp [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+  // Set received swing status
+  if (remote_state[0] & 0x02) {
+    ESP_LOGV(TAG, "Swing vertical");
+    this->swing_mode = climate::CLIMATE_SWING_VERTICAL;
+  } else {
+    ESP_LOGV(TAG, "Swing OFF");
+    this->swing_mode = climate::CLIMATE_SWING_OFF;
+  }
+
+  for (unsigned char &i : remote_state)
```
Comment (jesserockz) [inline esphome/components/noblex/noblex.cpp]:
```suggestion
  for (uint8_t &i : remote_state)
```

File: esphome/components/noblex/noblex.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+  // Footer
+  data->mark(NOBLEX_BIT_MARK);
+
+  transmit.perform();
+}  // end transmit_state()
+
+// Handle received IR Buffer
+bool NoblexClimate::on_receive(remote_base::RemoteReceiveData data) {
+  static bool receiving = false;
+  static uint8_t remote_state[8] = {0};
```
Comment (jesserockz) [inline esphome/components/noblex/noblex.cpp]: These should not be static variables, if you need them across calls, make them member variables please.

File: esphome/components/noblex/noblex.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  bool powered_on_assumed;
+
+ protected:
+  uint8_t swap_(uint8_t num);
+  /// Transmit via IR the state of this climate controller.
+  void transmit_state() override;
+  /// Handle received IR Buffer.
+  bool on_receive(remote_base::RemoteReceiveData data) override;
+
+  bool send_swing_cmd_{false};
```
Comment (jesserockz) [inline esphome/components/noblex/noblex.h:42]: Not actually used anywhere?

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/noblex/noblex.cpp (hunk 2 above)
Comment (AGalfra) [reply esphome/components/noblex/noblex.cpp]: made the suggested changes

File: esphome/components/noblex/noblex.cpp (hunk 3 above)
Comment (AGalfra) [reply esphome/components/noblex/noblex.cpp]: @jesserockz  made the suggested changes

File: esphome/components/noblex/noblex.cpp (hunk 4 above)
Comment (AGalfra) [reply esphome/components/noblex/noblex.cpp]: @jesserockz made the suggested changes

File: esphome/components/noblex/noblex.cpp (hunk 5 above)
Comment (AGalfra) [reply esphome/components/noblex/noblex.cpp]: revision of variables carried out

File: esphome/components/noblex/noblex.h (hunk 6 above)
Comment (AGalfra) [reply esphome/components/noblex/noblex.h:42]: @jesserockz  revision of variables carried out

Comment (AGalfra) [reply esphome/components/noblex/noblex.h:42]: It is the same as in Whirpool, LG, Coolix...

Comment (AGalfra) [reply esphome/components/noblex/noblex.h:42]: @jesserockz
Hello dear, how do you now find this integration so that you can merge it?

Comment (AGalfra) [reply esphome/components/noblex/noblex.h:42]: @jesserockz how do you now find this integration so that you can merge it?

**Review by jesserockz — APPROVED** [review]

**Pull request merged**